server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +150,62 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        try (TestThreadPool tp = new TestThreadPool(getTestName())) {
+            final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+            final int threads = 4;
+            final AtomicBoolean running = new AtomicBoolean(true);
+            final CountDownLatch startBarrier = new CountDownLatch(threads + 1);
```

Comment:
Nit: can we call this `startLatch` instead of `startBarrier`. The name got me for a bit.